contrib/native_window_msg/src/com/sun/jna/platform/win32/Win32WindowMessagesDemo.java

```diff
@@ -0,0 +1,367 @@
+ * Apache License 2.0. (starting with JNA version 4.0.0).
+ * 
+ * You can freely decide which license you want to apply to 
+ * the project.
+ * 
+ * You may obtain a copy of the LGPL License at:
+ * 
+ * http://www.gnu.org/licenses/licenses.html
+ * 
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "LGPL2.1".
+ * 
+ * You may obtain a copy of the Apache License at:
+ * 
+ * http://www.apache.org/licenses/
+ * 
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+
+package com.sun.jna.platform.win32;
+
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+
+import com.sun.jna.Native;
+import com.sun.jna.Pointer;
+import com.sun.jna.Structure;
+import com.sun.jna.platform.win32.BaseTSD.ULONG_PTR;
+import com.sun.jna.platform.win32.WinDef.DWORD;
+import com.sun.jna.platform.win32.WinDef.HINSTANCE;
+import com.sun.jna.platform.win32.WinDef.HMODULE;
+import com.sun.jna.platform.win32.WinDef.HWND;
+import com.sun.jna.platform.win32.WinDef.LPARAM;
+import com.sun.jna.platform.win32.WinDef.LRESULT;
+import com.sun.jna.platform.win32.WinDef.WPARAM;
+import com.sun.jna.platform.win32.WinUser.COPYDATASTRUCT;
+import com.sun.jna.platform.win32.WinUser.HHOOK;
+import com.sun.jna.platform.win32.WinUser.HOOKPROC;
+import com.sun.jna.platform.win32.WinUser.MSG;
+import com.sun.jna.platform.win32.WinUser.WNDCLASSEX;
+import com.sun.jna.platform.win32.WinUser.WNDENUMPROC;
+import com.sun.jna.platform.win32.WinUser.WindowProc;
+
+public class Win32WindowMessagesDemo {
+
+	public static final int MSG_CODE = 101;
+
+	public static final int DATA_STRUCT_CODE = 129;
+
+	public static final int MSG_CODE_HOOKED = 711;
+
+	/**
+	 * The main method.
+	 * 
+	 * @param args
+	 *            the arguments
+	 */
+	public static void main(String[] args) {
+		new Win32WindowMessagesDemo();
+	}
+
+	/**
+	 * Instantiates 2 windows and make them communicate through windows messages, even complex ones throught
+	 * WM_COPYDATA.
+	 */
+	public Win32WindowMessagesDemo() {
+		// Create window 1 named "ping"
+		createWindow("ping");
+
+		// let windows create the window before searching its handle.
+		sleepCurrThread(4000);
+		// Retrieves the created window's handle.
+		HWND hwndPing = determineHWNDFromWindowClass("ping");
+
+		// DEMO 1 : sends a simple message to ping window with code MSG_CODE and value 123456.
+		LRESULT result = User32.INSTANCE.SendMessage(hwndPing, WinUser.WM_USER, new WPARAM(MSG_CODE), new LPARAM(123456));
+		log("User Message sent to " + hwndPing + ", result = " + result);
+
+		// DEMO 2 : send of structured message.
+		COPYDATASTRUCT copyDataStruct = createStructuredMessage();
+		result = User32.INSTANCE.SendMessage(hwndPing, WinUser.WM_COPYDATA, null /* No current hwnd for this demo */,
+		        new LPARAM(copyDataStruct.getPointer().getLong(0)));
```

Comment:
the `copyDataStruct.getPointer().getLong(0)` looks wrong. I asume `Pointer.nativeValue(copyDataStruct.getPointer())` is what you mean. Also it might be worth mentioning in the comments, that copyDataStruct must be held strongly on the java side. In the sample this is archived by reading from the structure (line 92), but if that would not present, copyDataStruct becomes eligible for GC potentially to early. This leads to difficult to track down errors.